### ISSUE TITLE
Use compiler-native IDs instead of our `id`/`new-id`?

### ISSUE THREAD

**Issue by aartaka** [atlas-engineer/nyxt#2991]
Our ID indexes objects we create, but indexing requires either:
- Adding an `id` slot to an object.
- Or storing the ID-object pair somewhere else.

Both are cumbersome. What if we had some Lisp-native way to generate object IDs without any work on our side? Like... 
- `print-unreadable-object`'s `:identity` option
- or `sxhash`! 

Most implementations use some kind of implementation-internal object ID fetching for unreadable objects, and most have Lisp APIs for getting this ID reliably. Here's what I've been able to hack up for SBCL, CCL, ECL, and ABCL:
``` lisp
(defun id (object)
  #+sbcl (sb-kernel::get-lisp-obj-address object)
  #+ccl (ccl::%address-of object)
  ;; Not perfect, but ECL doesn't leave us much to do here.
  ;; Maybe use a plain `sxhash' instead?
  #+ecl (parse-integer
         (subseq (with-output-to-string (s)
                   (si:print-unreadable-object-function object s nil t (lambda ())))
                 5) ; To strip "#< 0x" off and read hex directly.
         :junk-allowed t
         :radix 16)
  #+abcl (system::identity-hash-code object)
  #-(or sbcl ccl ecl abcl) (sxhash object))
```

For all the other implementations there's always `sxhash` which has nice properties (`equal` objects have the same `sxhash`), but IDs are even better and come from the reliable compiler internals.

So, shall we go down that road?

**Comment by aadcg** [comment]
Sounds interesting to me.

**Comment by Ambrevar** [comment]
Very smart, I like it! :)

**Comment by aartaka** [comment]
Updated function for some more implementations I've tested:
``` lisp
(defun id (object)
  #+sbcl (sb-kernel::get-lisp-obj-address object)
  #+ccl (ccl::%address-of object)
  #+ecl (si:pointer object)
  #+abcl (system::identity-hash-code object)
  #+clisp (system::address-of object)
  #+gcl (system:address object)
  #-(or sbcl ccl ecl abcl clisp gcl) (sxhash object))
```

**Comment by Ambrevar** [comment]
Nonguix has AllegroCL if you feel like testing it :)

**Comment by aartaka** [comment]
> Nonguix has AllegroCL if you feel like testing it :)

The license has expired and I've sent a patch to Nonguix to fix it :P

**Comment by jmercouris** [comment]
I don't see the benefit here. In fact if we serialize an object where we have the id in a slot, that id will be persisted. With this method, it would not, no?

**Comment by aartaka** [comment]
> I don't see the benefit here. In fact if we serialize an object where we have the id in a slot, that id will be persisted. With this method, it would not, no?

Yes, but wan can [compile the data into .FASL-s to preserve the object itself](https://github.com/atlas-engineer/nfiles/issues/5), for example.
